File: examples/csharp/_common/MarketData.cs
namespace Wickra.Examples;

/// <summary>One OHLCV bar with a millisecond timestamp.</summary>
public readonly record struct Bar(double Open, double High, double Low, double Close, double Volume, long Timestamp);

/// <summary>
/// Deterministic synthetic market data plus a small OHLCV CSV loader, shared by
/// the offline examples so they run without network access.
/// </summary>
public static class MarketData
{
    /// <summary>A reproducible price path (trend + two cycles), no randomness.</summary>
    public static double[] SyntheticPrices(int count, double start = 100.0)
    {
        var prices = new double[count];
        for (var i = 0; i < count; i++)
        {
            prices[i] = start + 12.0 * Math.Sin(i * 0.05) + 5.0 * Math.Sin(i * 0.013) + i * 0.01;
        }

        return prices;
    }

    /// <summary>A reproducible OHLCV series derived from <see cref="SyntheticPrices"/>.</summary>
    public static Bar[] SyntheticCandles(int count, long startTimestamp = 0, long stepMs = 3_600_000)
    {
        var prices = SyntheticPrices(count + 1);
        var bars = new Bar[count];
        for (var i = 0; i < count; i++)
        {
            var open = prices[i];
            var close = prices[i + 1];
            var high = Math.Max(open, close) + 0.5 + Math.Abs(Math.Sin(i * 0.7));
            var low = Math.Min(open, close) - 0.5 - Math.Abs(Math.Cos(i * 0.7));
            var volume = 1_000.0 + 500.0 * (1.0 + Math.Sin(i * 0.1));
            bars[i] = new Bar(open, high, low, close, volume, startTimestamp + i * stepMs);
        }

        return bars;
    }

    /// <summary>
    /// Loads an OHLCV CSV. Accepts rows of <c>timestamp,open,high,low,close,volume</c>
    /// or <c>open,high,low,close,volume</c>; a non-numeric first row is treated as a header.
    /// </summary>
    public static Bar[] LoadOhlcvCsv(string path)
    {
        var bars = new List<Bar>();
        foreach (var rawLine in File.ReadLines(path))
        {
            var line = rawLine.Trim();
            if (line.Length == 0)
            {
                continue;
            }

            var cols = line.Split(',');
            if (!double.TryParse(cols[0], System.Globalization.CultureInfo.InvariantCulture, out _) &&
                !long.TryParse(cols[0], out _))
            {
                continue; // header row
            }

            double F(int i) => double.Parse(cols[i], System.Globalization.CultureInfo.InvariantCulture);

            if (cols.Length >= 6)
            {
                bars.Add(new Bar(F(1), F(2), F(3), F(4), F(5), long.Parse(cols[0])));
            }
            else
            {
                bars.Add(new Bar(F(0), F(1), F(2), F(3), F(4), bars.Count));
            }
        }

        return bars.ToArray();
    }
}

File: examples/java/src/main/java/org/wickra/examples/MarketData.java
package org.wickra.examples;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/**
 * Deterministic synthetic market data plus a small OHLCV CSV loader, shared by
 * the offline examples so they run without network access.
 */
public final class MarketData {
    private MarketData() {
    }

    /** One OHLCV bar with a millisecond timestamp. */
    public record Bar(double open, double high, double low, double close, double volume, long timestamp) {
    }

    /** A reproducible price path (trend + two cycles), no randomness. */
    public static double[] syntheticPrices(int count) {
        return syntheticPrices(count, 100.0);
    }

    public static double[] syntheticPrices(int count, double start) {
        double[] prices = new double[count];
        for (int i = 0; i < count; i++) {
            prices[i] = start + 12.0 * Math.sin(i * 0.05) + 5.0 * Math.sin(i * 0.013) + i * 0.01;
        }
        return prices;
    }

    /** A reproducible OHLCV series derived from {@link #syntheticPrices(int)}. */
    public static Bar[] syntheticCandles(int count) {
        return syntheticCandles(count, 0L, 3_600_000L);
    }

    public static Bar[] syntheticCandles(int count, long startTimestamp, long stepMs) {
        double[] prices = syntheticPrices(count + 1);
        Bar[] bars = new Bar[count];
        for (int i = 0; i < count; i++) {
            double open = prices[i];
            double close = prices[i + 1];
            double high = Math.max(open, close) + 0.5 + Math.abs(Math.sin(i * 0.7));
            double low = Math.min(open, close) - 0.5 - Math.abs(Math.cos(i * 0.7));
            double volume = 1_000.0 + 500.0 * (1.0 + Math.sin(i * 0.1));
            bars[i] = new Bar(open, high, low, close, volume, startTimestamp + (long) i * stepMs);
        }
        return bars;
    }

    /**
     * Loads an OHLCV CSV. Accepts rows of {@code timestamp,open,high,low,close,volume}
     * or {@code open,high,low,close,volume}; a non-numeric first row is treated as a header.
     */
    public static Bar[] loadOhlcvCsv(String path) throws IOException {
        List<Bar> bars = new ArrayList<>();
        for (String rawLine : Files.readAllLines(Path.of(path))) {
            String line = rawLine.trim();
            if (line.isEmpty()) {
                continue;
            }
            String[] cols = line.split(",");
            if (!isNumeric(cols[0])) {
                continue; // header row
            }
            if (cols.length >= 6) {
                bars.add(new Bar(
                        Double.parseDouble(cols[1]), Double.parseDouble(cols[2]),
                        Double.parseDouble(cols[3]), Double.parseDouble(cols[4]),
                        Double.parseDouble(cols[5]), Long.parseLong(cols[0])));
            } else {
                bars.add(new Bar(
                        Double.parseDouble(cols[0]), Double.parseDouble(cols[1]),
                        Double.parseDouble(cols[2]), Double.parseDouble(cols[3]),
                        Double.parseDouble(cols[4]), bars.size()));
            }
        }
        return bars.toArray(new Bar[0]);
    }

    private static boolean isNumeric(String s) {
        try {
            Double.parseDouble(s);
            return true;
        } catch (NumberFormatException e) {
            return false;
        }
    }
}
